domain/state.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package domain
+
+import (
+	"sync"
+
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+)
+
+// DBFactory defines a function that returns a database or an error if a
+// database cannot be returned.
+type DBFactory = func() (database.TrackedDB, error)
+
+// NamespaceDBFactory returns a DBFactory that returns a database from the
+// given getter function.
+func NamespaceDBFactory(getter func(string) (database.TrackedDB, error), namespace string) DBFactory {
```

Comment:
Not convinced we'll need this. Can we drop it from this patch?

Reply:
Sure.